src/provider/model/AppDirectory.ts

```diff
@@ -97,54 +73,69 @@ export class AppDirectory extends AsyncInit {
-        if (process.env.NODE_ENV === 'production') {
-            await this.updateDirectory([]);
-            return;
+    private async initializeDirectoryData(): Promise<void> {
+        this._url = this._configStore.config.query({level: 'desktop'}).applicationDirectory;
+        const fetchedData = await this.fetchOnline(this._url);
+        const cachedData = this.fetchCacheData();
+
+        if (fetchedData) {
+            this.updateUrl(this._url);
+            this.updateDirectory(fetchedData);
+        } else if (cachedData) {
+            this.updateDirectory(cachedData);
+        } else {
+            this.updateDirectory([]);
         }
-        const storedUrl = localStorage.getItem(StorageKeys.URL);
-        const currentUrl = this._url;
-        const applications = await this.fetchData(currentUrl, storedUrl);
-        await this.updateDirectory(applications);
-        await this.updateUrl(currentUrl);
     }
 
-    /**
-     * Fetch the AppDirectory.
-     * @param url Location of the AppDirectory.
-     * @param storedUrl Cache url
-     */
-    private async fetchData(url: string, storedUrl: string | null): Promise<Application[]> {
+    private async fetchOnline(url: string): Promise<Application[]|null> {
         const response = await fetch(url).catch(() => {
             console.warn(`Failed to fetch app directory @ ${url}`);
         });
 
         if (response && response.ok) {
-            return response.json();
+            try {
+                // validate the response is actually JSON
+                const validate = await response.json();
+                return validate;
+            } catch (error) {
+                console.warn(`Received invalid JSON data from ${url}. Ignoring fetch result`);
+            }
         }
-        // Use cached apps if urls match
-        if (url === storedUrl) {
-            return JSON.parse(localStorage.getItem(StorageKeys.APPLICATIONS) || '[]');
+
+        return null;
+    }
+
+    private fetchCacheData(): Application[]|null {
+        if (localStorage.getItem(StorageKeys.URL) === this._url) {
+            const cache = localStorage.getItem(StorageKeys.APPLICATIONS);
+
+            if (cache) {
+                try {
+                    const validate = JSON.parse(cache);
+                    return validate;
+                } catch (error) {
+                    // Not likely to get here but figured it's better to safely to handle it.
+                    console.warn('Invalid JSON retrieved from cache');
+                }
+            }
         }
-        // Use empty array if urls dont match
-        return [];
+
+        return null;
     }
 
     /**
      * Update the application directory in memory and storage.
      * @param applications To place into the directory.
      */
-    private async updateDirectory(applications: Application[]): Promise<void> {
+    private updateDirectory(applications: Application[]): void {
```

Comment:
I'm not sure we want to do that.  These are only used in one place. We need to only update the URL if the fetch succeeds, otherwise we would lose the cache if the application were to be reloaded as we check the URL provided from the application directory against the one in storage.

Reply:
How about
    if (fetchedData)
    {
        this.updateCache(this._url, fetchedData);
    }
    this.updateDirectory(fetchedData || cachedData || []);
?

Or update the cache from within `fetchOnline` instead of `updateDirectory`?

Reply:
I still feel that is not much better as we are now writing the cache twice in the case of truthy fetchedData (which I noticed is also the case with calling it with cachedData...) and we are just replacing one method with another.  

Would you settle with: 
```
        if (fetchedData) {
            this.updateUrl(this._url);
            this.updateDirectory(fetchedData);
        } else {
            this._directory = cachedData || []
        }

```
Or keep the above structure and remove the methods completely, and update the members / call storage directly in the if?

Reply:
Let's talk about this one, grab me when you've got two minutes

Reply:
Fixed as discussed